src/symbolic_memory.ml

```diff
@@ -71,6 +71,14 @@ module M = struct
 
   let concat ~msb ~lsb offset =
     assert (offset > 0 && offset <= 8);
+    let merge_extracts (e1, h, m1) (e2, m2, l) =
+      if not (m1 = m2 && Expr.equal e1 e2) then
+        Expr.(Extract (e1, h, m1) ++ Extract (e2, m2, l))
+      else
+        match Expr.type_of e1 with
+        | Some t -> if h - l = Types.size t then e1 else Extract (e1, h, l)
+        | None -> Extract (e1, h, l)
+    in
```

Comment:
This can be written a little bit shorter:

```ocaml
        match Expr.type_of e1 with
        | Some t when h - l = Types.size t -> e1
        | None | Some _ -> Extract (e1, h, l)
```